File: tools/src/ilp_global_solver/ilp_global_solver.py
#! /usr/bin/python3
from common import arg_parser as ap
from common import instance
from typing import List, Tuple
import gurobipy as grb
import sys
import time


class Solver:

    def __init__(self, arg_parser: ap.ArgParser, env: instance.Environment, acs: List[instance.AssignmentCharacteristic], timelimit:float=float("inf")):
        self.arg_parser = arg_parser
        self.env = env
        self.acs = acs        
        self.timelimit = timelimit
        self.model = None
        self.x_ijk = None
        self.l_j = None
        
        self._init()
        
    def _init(self):
        num_tasks = len(self.acs)
        windows_ub = num_tasks
        env = self.env

        model = grb.Model()
        model.setParam("TimeLimit", self.timelimit)

        # VARIABLES
        # x_ijk variables        
        x_ijk = model.addVars([(i, j, k)
                               for i in range(num_tasks)
                               for j in range(windows_ub)
                               for k in range(len(self.acs[i].resource_assignmnets))],
                              vtype=grb.GRB.BINARY,
                              name="x")        

        # l_j variables
        l_j = model.addVars(windows_ub, vtype=grb.GRB.INTEGER, name="l")

        # B_j variables to linearize max b_ijk
        B_j = model.addVars(windows_ub, vtype=grb.GRB.CONTINUOUS, lb=0, name="B")

        # CONSTRAINTS
        # - assignment constraint
        model.addConstrs(x_ijk.sum(i, "*", "*") == 1
                         for i in range(num_tasks))
        
        # - window length constraint - batching
        model.addConstrs(l_j[j] >= ((x_ijk[i, j, k] * self.acs[i].resource_assignmnets[k].length) / env.sc_part)
                         for i in range(num_tasks)
                         for j in range(windows_ub)
                         for k in range(len(self.acs[i].resource_assignmnets)))

        # - window length constraint - if nothing is assigned, window is empty                                
        model.addConstrs((l_j[j] <= env.major_frame_length * x_ijk.sum("*", j, "*")
                          for j in range(windows_ub)), name="window is not empty")

        # - window ordering constraint (symmetry pruning)
        model.addConstrs((l_j[j] >= l_j[j+1]
                         for j in range(windows_ub-1)), name="window ordering")

        # - major length constraint
        model.addConstr(l_j.sum("*") <= env.major_frame_length,
                        name="major frame length")

        # - resource capacity constraint
        model.addConstrs((grb.quicksum(x_ijk[i, j, k] * acp.processing_units
                                       for i in range(num_tasks)
                                       for k in range(len(self.acs[i].resource_assignmnets))
                                       for acp in self.acs[i].resource_assignmnets[k].processors if acp.processor == processor.name)
                          <= processor.processing_units
                         for processor in env.processors_list
                         for j in range(windows_ub)), name="resource capacity")

        # - link B_j with  b_ik coefficients        
        M = env.major_frame_length * max([self.acs[i].resource_assignmnets[k].intercept
                                            for i in range(num_tasks)
                                            for k in range(len(self.acs[i].resource_assignmnets))])
        model.addConstrs((B_j[j] >= env.sc_part * l_j[j] * self.acs[i].resource_assignmnets[k].intercept - M * (1-x_ijk[i, j, k])
                          for i in range(num_tasks)
                          for j in range(windows_ub)
                          for k in range(len(self.acs[i].resource_assignmnets))), name="B_j value link")

        # OBJECTIVE
        energy_consumption_sum=grb.quicksum(B_j[j] + grb.quicksum(x_ijk[i, j, k]
                                                                  * self.acs[i].resource_assignmnets[k].length
                                                                  * self.acs[i].resource_assignmnets[k].slope
                                                                  for i in range(num_tasks)
                                                                  for k in range(len(self.acs[i].resource_assignmnets)))
                                            for j in range(windows_ub))
        model.setObjectiveN(energy_consumption_sum * 1/env.major_frame_length,
                            index=0, priority=1, weight=1.0, name="min avg power consumption")

        if (self.arg_parser.is_arg_present("--optimize-schedule")):
            print("warning: --optimize-schedule active with predictor method", file=sys.stderr)
            model.setObjectiveN(l_j.sum("*"),
                                index=1, priority=0, weight=1.0, name="min total schedule length")
            
        self.model = model
        self.x_ijk = x_ijk
        self.l_j = l_j

    def solve(self) -> Tuple[instance.Solution, List[instance.Task]]:
        num_tasks = len(self.acs)
        windows_ub = num_tasks
        env = self.env
        model = self.model
        x_ijk = self.x_ijk
        l_j = self.l_j

        # OPTIMIZE THE MODEL
        t_start=time.time()
        model.optimize()
        t_end=time.time()

        solution=None
        tasks=[]

        s_feasible=True if model.Status == grb.GRB.OPTIMAL or (
            model.Status == grb.GRB.TIME_LIMIT and model.SolCount > 0) else False
        s_solver_name="ILP Solver (global):predictor (Python)"
        s_solution_time=int(round((t_end - t_start)*1000))  # to ms
        s_windows=[]
        s_metadata={"objective": str(float("inf"))}

        if s_feasible:
            s_metadata["objective"] = str(model.ObjVal)
            for j in range(windows_ub):
                window_length=int(round(l_j[j].X))
                window_tasks_assignments=[]
                if window_length < 1:
                    continue

                # processor-unit allocation (start with all processors empty)
                pu_allocations={p.name: 0 for p in env.processors_list}

                for i in range(num_tasks):
                    push_task=False
                    task_characteristic=self.acs[i]
                    task_processors=[]
                    task_length=-1

                    for k, ac in enumerate(task_characteristic.resource_assignmnets):
                        if x_ijk[i, j, k].X > 0.5:
                            task_ass_idx = k
                            push_task=True   
                            for p in ac.processors:
                                window_tasks_assignments.append(instance.TaskAssignment(task=task_characteristic.task,
                                                                                        processor=p.processor,
                                                                                        processing_unit=pu_allocations[p.processor],
                                                                                        start=0,
                                                                                        length=ac.length))

                                # increment the processor units
                                pu_allocations[p.processor] += p.processing_units

                                task_processors.append(instance.ProcessorAssignment(
                                    p.processor, p.processing_units))
                                task_length=ac.length


                    if push_task:
                        tasks.append(instance.Task(name=task_characteristic.task,
                                                   command=task_characteristic.command,
                                                   length=task_length,
                                                   assignment_index=task_ass_idx,
                                                   processors=task_processors))
            
                s_windows.append(instance.Window(window_length, window_tasks_assignments))
        else:  # infeasible solution            
            s_feasible=False
            if self.arg_parser.is_arg_present("--iis-output"):
                model.computeIIS()
                model.write(self.arg_parser.get_arg_value("--iis-output") + ".ilp")        
        solution = instance.Solution(s_feasible, s_solver_name, s_solution_time, s_metadata, s_windows)
        
        return (solution, tasks)


class SolverFixed:
    def __init__(self, arg_parser: ap.ArgParser, env: instance.Environment, acs: List[instance.AssignmentCharacteristic], timelimit:float=float("inf"), cutoff=None):
        self.arg_parser = arg_parser
        self.env = env
        self.acs = acs                
        self.timelimit = timelimit
        self.cutoff=cutoff
        
        self.model = None
        self.a_ikln = None
        self.x_ln = None
        self.B_ln = None
        self.task_lengths = None
        self.task_lengths_num = None
        self.length_to_task_idx_and_conf = None
        self.task_idx_to_possible_lengths = None        
        
        self._init()          
        
    def _init(self):
        task_to_idx = {}
        for i, a in enumerate(self.acs):
            task_to_idx[a.task] = i
        num_tasks = len(self.acs)
            
        m = grb.Model("GlobalModelIlpFixed")    
        m.setParam("TimeLimit", self.timelimit)            
        
        # prepare some structures:
        task_lengths, task_lengths_num = instance.get_tasks_lengths_and_nums(self.acs)
        length_to_task_idx_and_conf = instance.get_length_to_task_conf(self.acs, task_lengths)        
        task_idx_to_possible_lengths = instance.get_task_conf_to_possible_lengths(self.acs, task_lengths)
        
        # VARIABLES
        a_ikln = m.addVars([(i, k, l,n)
                           for i in range(num_tasks)                         
                           for k in range(len(self.acs[i].resource_assignmnets))
                           for l in task_idx_to_possible_lengths[(i,k)]
                           for n in range(task_lengths_num[l])],
                          vtype=grb.GRB.BINARY,
                          name="a")  
        x_ln = m.addVars([(l,n) for l in task_lengths for n in range(task_lengths_num[l])], vtype=grb.GRB.BINARY, name="x")
        B_ln = m.addVars([(l,n) for l in task_lengths for n in range(task_lengths_num[l])], vtype=grb.GRB.CONTINUOUS, name="B")
        
        # CONSTRAINTS
        # - link x and a (if assigned then x := 1)
        M = sum([p.processing_units for p in self.env.processors_list])
        m.addConstrs(M*x_ln[l,n] >= a_ikln.sum("*", "*", l,n) for l in task_lengths for n in range(task_lengths_num[l]))
        # - if window is selected, task with appropriate length is assigned there
        m.addConstrs(x_ln[l,n] <= grb.quicksum(a_ikln[ti, tk, l,n] for ti, tk in length_to_task_idx_and_conf[l]) for l in task_lengths for n in range(task_lengths_num[l]))
        # - windows need to fit the MF
        m.addConstr(grb.quicksum(x_ln[l,n] * l for l in task_lengths for n in range(task_lengths_num[l])) <= self.env.major_frame_length)
        # - every task is scheduled
        m.addConstrs(a_ikln.sum(i,"*", "*", "*") == 1 for i in range(num_tasks))
        # - resource capacities
        m.addConstrs((grb.quicksum(a_ikln[i,k,l,n] * acp.processing_units
                                   for i in range(num_tasks)
                                   for k in range(len(self.acs[i].resource_assignmnets))
                                   for acp in self.acs[i].resource_assignmnets[k].processors if acp.processor == processor.name
                                   if l in task_idx_to_possible_lengths[(i,k)])
                      <= processor.processing_units)
                      for processor in self.env.processors_list
                      for l in task_lengths
                      for n in range(task_lengths_num[l]))
        
        # offset coefficient
        m.addConstrs(B_ln[l,n] >= a_ikln[i,k,l,n]*self.acs[i].resource_assignmnets[k].intercept for l in task_lengths for n in range(task_lengths_num[l]) for i in range(num_tasks) for k in range(len(self.acs[i].resource_assignmnets)) if l in task_idx_to_possible_lengths[(i,k)])
        
        # objective
        task_cost = grb.quicksum(a_ikln.sum(i,k,"*","*") * self.acs[i].resource_assignmnets[k].slope*self.acs[i].resource_assignmnets[k].length for i in range(num_tasks) for k in range(len(self.acs[i].resource_assignmnets)) )
        m.setObjective((grb.quicksum(B_ln[l,n]*l for l in task_lengths for n in range(task_lengths_num[l])) + task_cost)*(1/self.env.major_frame_length))
       
        
        self.model = m
        self.a_ikln = a_ikln
        self.x_ln = x_ln
        self.B_ln = B_ln
        self.task_lengths = task_lengths
        self.task_lengths_num = task_lengths_num
        self.length_to_task_idx_and_conf = length_to_task_idx_and_conf
        self.task_idx_to_possible_lengths = task_idx_to_possible_lengths

    def solve(self) -> Tuple[instance.Solution, List[instance.Task]]:
        num_tasks = len(self.acs)        
        env = self.env
        model = self.model
        a_ikln = self.a_ikln
        x_ln = self.x_ln

        # OPTIMIZE THE MODEL
        t_start=time.time()
        model.optimize()
        t_end=time.time()        

        solution=None
        tasks=[]

        s_feasible=True if model.Status == grb.GRB.OPTIMAL or (
            model.Status == grb.GRB.TIME_LIMIT and model.SolCount > 0) else False
        s_solver_name="ILP-Global2"
        s_solution_time=int(round((t_end - t_start)*1000))  # to ms
        s_windows=[]
        s_metadata={"objective": str(float("inf"))}

        if s_feasible:
            s_metadata["objective"] = str(model.ObjVal)
            
            for l in self.task_lengths:
                for n in range(self.task_lengths_num[l]):
                    if self.x_ln[l,n].X > 0.5:
                        window_length = l
                        window_tasks_assignments=[]
                        
                        # processor-unit allocation (start with all processors empty)
                        pu_allocations={p.name: 0 for p in env.processors_list}
                        
                        for i in range(num_tasks):  
                            for k in range(len(self.acs[i].resource_assignmnets)):
                                if l in self.task_idx_to_possible_lengths[(i,k)] and a_ikln[i,k,l,n].X > 0.5:
                                    task_characteristic=self.acs[i]                                
                                    ac = task_characteristic.resource_assignmnets[k]
                                    task_processors=[]
                                    task_length=ac.length
                                                        
                                    for p in ac.processors:
                                        window_tasks_assignments.append(instance.TaskAssignment(task=task_characteristic.task,
                                                                                                processor=p.processor,
                                                                                                processing_unit=pu_allocations[p.processor],
                                                                                                start=0,
                                                                                                length=task_length))

                                        # increment the processor units
                                        pu_allocations[p.processor] += p.processing_units
                                        task_processors.append(instance.ProcessorAssignment(p.processor, p.processing_units))                            


                                
                                    tasks.append(instance.Task(name=task_characteristic.task,
                                                            command=task_characteristic.command,
                                                            length=task_length,
                                                            assignment_index=k,
                                                            processors=task_processors))
                    
                        s_windows.append(instance.Window(window_length, window_tasks_assignments))
                                                           
        else:  # infeasible solution            
            s_feasible=False
            if self.arg_parser.is_arg_present("--iis-output"):
                model.computeIIS()
                model.write(self.arg_parser.get_arg_value("--iis-output") + ".ilp")        
        solution = instance.Solution(s_feasible, s_solver_name, s_solution_time, s_metadata, s_windows)
        
        return (solution, tasks)        

File: tools/src/ilp_global_solver/ilp_global_solver.cpp
#include <iostream>
#include <chrono>
#include <gurobi_c++.h>

#include "../common/instance.h"
#include "../common/arg_parser.h"
#include "../common/ilp_util.h"
#include "ilp_global_solver.h"

#define INFEASIBLE_MODEL_FILENAME std::string("infeasible_global_problem")

// old ILP
std::pair<solution, std::vector<task>> solve_eik(const arg_parser& args, const environment& e, const assignment_characteristic_list& assignment_characteristics)
{
	// leftover model from development
	std::vector<std::vector<std::vector<GRBVar>>> a;

	GRBEnv env;
	GRBModel model(env);

	int num_tasks = assignment_characteristics.size();
	int windows_ub = num_tasks;
	int sense = 1;
	if (args.is_arg_present("--maximize"))
		sense = -1;

	GRBLinExpr energy_consumption_sum;
	for (int i = 0; i < num_tasks; i++)
	{
		auto& task = assignment_characteristics[i];
		std::vector<std::vector<GRBVar>> a_i;
		GRBLinExpr assignment_constraint_expr;

		for (int j = 0; j < windows_ub; j++)
		{
			std::vector<GRBVar> a_ij;

			// k
			for (auto& assignment_characteristic : task.resource_assignments)
			{
				GRBVar a_ijk = model.addVar(0, 1, 0, GRB_BINARY, "a" + std::to_string(i) + "," + std::to_string(j) + "," + std::to_string(a_ij.size()));

				int processor_capacity = 0;
				for (auto& p : assignment_characteristic.processors)
					processor_capacity += e.processors.at(p.processor).processing_units;
				float energy_consumption = assignment_characteristic.slope + (float) assignment_characteristic.intercept / (float) processor_capacity;
				energy_consumption_sum += a_ijk * energy_consumption * assignment_characteristic.length * sense;
				assignment_constraint_expr += a_ijk;
				a_ij.push_back(std::move(a_ijk));
			}

			a_i.push_back(std::move(a_ij));
		}

		model.addConstr(assignment_constraint_expr == 1, task.task + " assigned");
		a.push_back(std::move(a_i));
	}

	std::vector<GRBVar> l(windows_ub);
	GRBLinExpr window_length_sum;
	for (int j = 0; j < windows_ub; j++)
	{
		l[j] = model.addVar(0, GRB_INFINITY, 0, GRB_INTEGER, "l" + std::to_string(j));

		window_length_sum += l[j];

		GRBLinExpr a_ik_sum;
		for (int i = 0; i < num_tasks; i++)
		{
			auto& task = assignment_characteristics[i];
			int k = 0;
			for (auto& assignment_characteristic : task.resource_assignments)
			{
				a_ik_sum += a[i][j][k];
				model.addConstr(l[j] >= (a[i][j][k++] * assignment_characteristic.length) / e.sc_part, task.task + " is at most " + std::to_string(e.sc_part*100) + "% of l" + std::to_string(j));
			}
		}
		model.addConstr(l[j] <= a_ik_sum * e.major_frame_length, "window " + std::to_string(j) + " is not empty");

		if (j > 0)
			model.addConstr(l[j] <= l[j - 1], "window ordering");
	}
	model.addConstr(window_length_sum <= e.major_frame_length, "major frame length");

	// constraint 5(+6)
	for (auto& processor : e.processors_list)
	{
		for (int j = 0; j < windows_ub; j++)
		{
			GRBLinExpr resource_capacity_contraint_expr;

			for (int i = 0; i < num_tasks; i++)
			{
				auto& task = assignment_characteristics[i];

				int k = 0;
				for (auto& assignment_characteristic : task.resource_assignments)
				{
					for (auto& acp : assignment_characteristic.processors)
					{
						if (acp.processor == processor->name)
						{
							resource_capacity_contraint_expr += a[i][j][k] * acp.processing_units;
						}
					}
					k++;
				}
			}

			model.addConstr(resource_capacity_contraint_expr <= processor->processing_units, processor->name + " capacity (window " + std::to_string(j) + ")");
		}
	}

	model.setObjectiveN(energy_consumption_sum, 0, 1, 1.0, 0.0, 0.0, "min energy consumption");

	if (args.is_arg_present("--optimize-schedule"))
		model.setObjectiveN(window_length_sum, 1, 0, 1.0, 0.0, 0.0, "min total schedule length");

	auto start = std::chrono::high_resolution_clock::now();
	model.optimize();
	auto end = std::chrono::high_resolution_clock::now();

	solution s{ .solver_name = "ILP Solver (global):eik",
			.solution_time = (uint64_t)std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() };
	std::vector<task> tasks;

	if (model.get(GRB_IntAttr_Status) == GRB_OPTIMAL || model.get(GRB_IntAttr_Status) == GRB_TIME_LIMIT)
	{
		s.feasible = true;

		for (int j = 0; j < windows_ub; j++)
		{
			int window_length = (int)l[j].get(GRB_DoubleAttr_X);
			if (window_length < 1)
				continue;

			std::unordered_map<std::string, int> pu_allocations;
			window win{ .length = window_length };

			for (int i = 0; i < num_tasks; i++)
			{
				auto& task_characteristic = assignment_characteristics[i];
				task task_definition{ .name = task_characteristic.task, .command = task_characteristic.command };
				bool push_task = false;

				int k = 0, aix = 0;
				for (auto& assignment_characteristic : task_characteristic.resource_assignments)
				{
					if (a[i][j][k++].get(GRB_DoubleAttr_X) > 0.5)
					{
						for (auto& p : assignment_characteristic.processors)
						{
							win.task_assignments.push_back({ .task = task_characteristic.task,
																   .processor = p.processor,
																   .processing_unit = pu_allocations[p.processor],
																   .start = 0, .length = assignment_characteristic.length });
							pu_allocations[p.processor] = pu_allocations[p.processor] + p.processing_units;

							push_task = true;
							task_definition.processors.push_back({ .processor = p.processor, .processing_units = p.processing_units });
							task_definition.length = assignment_characteristic.length;
							task_definition.assignment_index = aix;
						}
					}
					aix++;
				}

				if (push_task)
					tasks.push_back(std::move(task_definition));
			}

			s.windows.push_back(std::move(win));
		}
	}
	else
	{
		s.feasible = false;
		bool iis_output = args.is_arg_present("--iis-output");
		if (iis_output)
		{
			model.computeIIS();
			write_iis(INFEASIBLE_MODEL_FILENAME, model);
		}
	}

	return std::make_pair(std::move(s), std::move(tasks));
}

// global-ILP
std::pair<solution, std::vector<task>> solve_global_ILP(const arg_parser& args, const environment& e, const assignment_characteristic_list& assignment_characteristics)
{
	// global-ILP model model
	int num_tasks = assignment_characteristics.size();
	int windows_ub = num_tasks;
	int sense = 1;
	if (args.is_arg_present("--maximize"))
		sense = -1;

	std::vector<std::vector<std::vector<GRBVar>>> x;

	GRBEnv env;
	GRBModel model(env);

	// x_ijk variables
	for (int i = 0; i < num_tasks; i++)
	{
		auto& task = assignment_characteristics[i];
		std::vector<std::vector<GRBVar>> x_i;
		GRBLinExpr assignment_constraint_expr;

		for (int j = 0; j < windows_ub; j++)
		{
			std::vector<GRBVar> x_ij;

			// task assignment constraint
			for (auto& assignment_characteristic : task.resource_assignments)
			{
				GRBVar x_ijk = model.addVar(0, 1, 0, GRB_BINARY, "x" + std::to_string(i) + "," + std::to_string(j) + "," + std::to_string(x_ij.size()));
				assignment_constraint_expr += x_ijk;
				x_ij.push_back(std::move(x_ijk));
			}

			x_i.push_back(std::move(x_ij));
		}

		model.addConstr(assignment_constraint_expr == 1, task.task + " assigned");
		x.push_back(std::move(x_i));
	}

	// l_j variables
	std::vector<GRBVar> l(windows_ub);
	GRBLinExpr window_length_sum;
	for (int j = 0; j < windows_ub; j++)
	{
		l[j] = model.addVar(0, GRB_INFINITY, 0, GRB_INTEGER, "l" + std::to_string(j));

		window_length_sum += l[j];

		GRBLinExpr a_ik_sum;
		for (int i = 0; i < num_tasks; i++)
		{
			auto& task = assignment_characteristics[i];
			int k = 0;
			for (auto& assignment_characteristic : task.resource_assignments)
			{
				a_ik_sum += x[i][j][k];
				// sc part constraint
				model.addConstr(l[j] >= (x[i][j][k++] * assignment_characteristic.length) / e.sc_part, task.task + " is at most " + std::to_string(e.sc_part*100) + "% of l" + std::to_string(j));
			}
		}

		// window length constraint
		model.addConstr(l[j] <= a_ik_sum * e.major_frame_length, "window " + std::to_string(j) + " is not empty");

		// window ordering constraint
		if (j > 0)
			model.addConstr(l[j] <= l[j - 1], "window ordering");
	}
	// major frame length constraint
	model.addConstr(window_length_sum <= e.major_frame_length, "major frame length");

	// A_ijk variables (=x_ijk*p_ik*a_ik), objective function
	GRBLinExpr energy_consumption_sum;
	for (int i = 0; i < num_tasks; i++)
	{
		auto& task = assignment_characteristics[i];

		for (int j = 0; j < windows_ub; j++)
		{
			int k = 0;
			for (auto& assignment_characteristic : task.resource_assignments)
			{
				auto& x_ijk = x[i][j][k];
				GRBVar A_ijk = model.addVar(0, GRB_INFINITY, 0, GRB_CONTINUOUS, "A" + std::to_string(i) + "," + std::to_string(j) + "," + std::to_string(k));
				model.addConstr(A_ijk == x_ijk * assignment_characteristic.slope * assignment_characteristic.length);
				energy_consumption_sum += A_ijk;
				k++;
			}
		}
	}

	// resource capacity constraint
	for (auto& processor : e.processors_list)
	{
		for (int j = 0; j < windows_ub; j++)
		{
			GRBLinExpr resource_capacity_contraint_expr;

			for (int i = 0; i < num_tasks; i++)
			{
				auto& task = assignment_characteristics[i];

				int k = 0;
				for (auto& assignment_characteristic : task.resource_assignments)
				{
					for (auto& acp : assignment_characteristic.processors)
					{
						if (acp.processor == processor->name)
						{
							resource_capacity_contraint_expr += x[i][j][k] * acp.processing_units;
						}
					}
					k++;
				}
			}

			model.addConstr(resource_capacity_contraint_expr <= processor->processing_units, processor->name + " capacity (window " + std::to_string(j) + ")");
		}
	}

	// B_ijk variables
	for (int j = 0; j < windows_ub; j++)
	{
		std::vector<GRBVar> B_j;
		for (auto& processor : e.processors_list)
		{
			for (int i = 0; i < num_tasks; i++)
			{
				auto& task = assignment_characteristics[i];
				int k = 0;
				for (auto& assignment_characteristic : task.resource_assignments)
				{
					for (auto& acp : assignment_characteristic.processors)
					{
						if (acp.processor == processor->name)
						{

							GRBVar B_ijk = model.addVar(0, GRB_INFINITY, 0, GRB_CONTINUOUS, "B_" + std::to_string(i) + "," + std::to_string(j) + "," + processor->name);

							model.addGenConstrIndicator(x[i][j][k], 1, B_ijk == e.sc_part * assignment_characteristic.intercept * l[j], "B_ijkVALUE(" + std::to_string(i) + "," + std::to_string(j) + "," + processor->name + ")");
							if (sense == -1)
								model.addGenConstrIndicator(x[i][j][k], 0, B_ijk == 0, "B_ijkVALUE(" + std::to_string(i) + "," + std::to_string(j) + "," + processor->name + ")_BOUND");
							B_j.push_back(std::move(B_ijk));
						}
					}
					k++;
				}
			}
		}

		GRBVar B_j_max = model.addVar(0, GRB_INFINITY, 0, GRB_CONTINUOUS, "B_" + std::to_string(j) + "_max");
		model.addGenConstrMax(B_j_max, B_j.data(), B_j.size(), 0, "B_" + std::to_string(j) + "_max_constr");
		energy_consumption_sum += B_j_max;
	}

	// minimize (default)/maximize objective
	if (sense == 1)
		model.setObjectiveN(energy_consumption_sum * 1.0f/(float) e.major_frame_length, 0, 1, 1.0, 0.0, 0.0, "min avg power consumption");
	else
		model.setObjective(energy_consumption_sum * 1.0f/(float) e.major_frame_length, GRB_MAXIMIZE);

	// enable C_max (schedule length) optimization (not recommended for global_ilp, leftover from development)
	if (args.is_arg_present("--optimize-schedule"))
	{
		if (sense == 1)
		{
			std::cerr << "warning: --optimize-schedule active with predictor method" << std::endl;
			model.setObjectiveN(window_length_sum, 1, 0, 1.0, 0.0, 0.0, "min total schedule length");
		}
		else if (sense == -1)
		{
			std::cerr << "error: --optimize-schedule is not compatible with --maximize while using predictor method, disregarding" << std::endl;
		}
	}

	// optimize the model
	auto start = std::chrono::high_resolution_clock::now();
	model.optimize();
	auto end = std::chrono::high_resolution_clock::now();

	solution s{ .solver_name = "ILP Solver (global):predictor",
			.solution_time = (uint64_t)std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() };
	std::vector<task> tasks;

	if (model.get(GRB_IntAttr_Status) == GRB_OPTIMAL || (model.get(GRB_IntAttr_Status) == GRB_TIME_LIMIT && model.get(GRB_IntAttr_SolCount) > 0))
	{
		// feasible solution
		s.feasible = true;

		for (int j = 0; j < windows_ub; j++)
		{
			int window_length = (int)l[j].get(GRB_DoubleAttr_X);
			if (window_length < 1)
				continue;

			std::unordered_map<std::string, int> pu_allocations;
			window win{ .length = window_length };

			for (int i = 0; i < num_tasks; i++)
			{
				auto& task_characteristic = assignment_characteristics[i];
				task task_definition{ .name = task_characteristic.task, .command = task_characteristic.command };
				bool push_task = false;

				int k = 0, aix = 0;
				for (auto& assignment_characteristic : task_characteristic.resource_assignments)
				{
					if (x[i][j][k++].get(GRB_DoubleAttr_X) > 0.5)
					{
						for (auto& p : assignment_characteristic.processors)
						{
							win.task_assignments.push_back({ .task = task_characteristic.task,
																   .processor = p.processor,
																   .processing_unit = pu_allocations[p.processor],
																   .start = 0, .length = assignment_characteristic.length });
							pu_allocations[p.processor] = pu_allocations[p.processor] + p.processing_units;

							push_task = true;
							task_definition.processors.push_back({ .processor = p.processor, .processing_units = p.processing_units });
							task_definition.length = assignment_characteristic.length;
							task_definition.assignment_index = aix;
						}
					}
					aix++;
				}

				if (push_task)
					tasks.push_back(std::move(task_definition));
			}

			s.windows.push_back(std::move(win));
		}
	}
	else
	{
		// infeasible solution
		s.feasible = false;
		bool iis_output = args.is_arg_present("--iis-output");
		if (iis_output)
		{
			model.computeIIS();
			write_iis(INFEASIBLE_MODEL_FILENAME, model);
		}
	}

	return std::make_pair(std::move(s), std::move(tasks));
}

std::pair<solution, std::vector<task>> solve(const arg_parser& args, const environment& e, const assignment_characteristic_list& assignment_characteristics)
{
	if (args.is_arg_present("--method"))
	{
		auto method = args.get_arg_value("--method");

		if (method == "global-ILP") return solve_global_ILP(args, e, assignment_characteristics);
		else if (method == "eik") return solve_eik(args, e, assignment_characteristics);
		else std::cerr << "optimization method not recognized, using default of global-ILP instead" << std::endl;
	}

	return solve_global_ILP(args, e, assignment_characteristics);
}